### DISCUSSION TITLE
about password error

### DISCUSSION THREAD

**Discussion by nenge123** [gildas-lormeau/zip.js#342] [Ideas]
if file need password but i use  errror password

i write a password in config,but zip not need password ,than has a PBKDFerr

i hope
------
I set a password in reader config, the reader first read file it not to using password to try read. if it real need a password ,then try to use password ,

password error : thow a password err message.

password true! then next file all using password.

add a try mode config
-----
all file are try to no password reader,when it err,try to using password to read,if err. then thow password error message.

![image](https://user-images.githubusercontent.com/16117315/178788599-c7df3a55-2e37-45f6-b746-a60f023d4ce4.jpeg)

**Comment by gildas-lormeau** [comment]
I confirm zip.js can work as you described. This is actually exactly how https://gildas-lormeau.github.io/zip.js/demos/demo-read-file.html detects and handle passwords. See also the code below. 
https://github.com/gildas-lormeau/zip.js/blob/9e65bce1557483154fe576319f4340898a858e93/tests/all/test-crypto.js#L21-L40

However I see that you're using "DZXF" so I guess the issue is in this library.

**Reply by nenge123** [reply]
this error is i using undefined password to read a no password file.


php(not working)
header('content-password','12345678');

js
passsword:headers['content-password']

file
m.zip (three empty txt file,and the file name   have UTF8/GBK)

error code
 let u8 = Entries.getData();

-------------

if i fix php not working,using a password to read a no-password zip

 let u8 = Entries.getData();
out put file "m.zip" same as source file

**Comment by gildas-lormeau** [comment]
Are you sure the call of `DZXF.unitl.unFile` is correct? What is the second parameter for example? In your code, I see a mysterious empty string (see `''`).
Moreover zip.js uses an **object** to pass options, e.g. `{ password: "kk" }` and not simply `'kk'` (see third parameter `'kk'` in your code) but again I have absolutely no clue about what `DZXF.unitl.unFile` is. Is this library documented somewhere?

**Comment by nenge123** [comment]
DZXF.unitl.unFile is my read zip/rar/7z

 DZXF.unitl.unFile(u8,callback,password,key)

check u8 head match func


       async unZip (u8,cb,password,key){
            let T=this.T,url = await this.getLibjs('zip.min.js');
            if(url instanceof File)url = this.URL(url);
            return new Promise(complete=>{
                let func = async e=>{
                    const reader = new zip.ZipReader(new zip.Uint8ArrayReader(u8),｛password｝);
                    const entries = await reader.getEntries();
                    if(entries.length>0){
                        let contents = {};
                        for(var i=0;i<entries.length;i++){
                            let file = entries[i];
                            if(!file.directory){
                                let u8,fileName=file.filename.split('/').pop();
                                u8 = await file.getData(new zip.Uint8ArrayWriter());
                                /*if(key){
                                    let mime = fileName.split('.').slice(-1);
                                    await T.setItem(
                                        T.LibStore,
                                        'libdata-'+fileName,
                                        {
                                            'contents':this.istext(mime)?new TextDecoder().decode(u8):u8,
                                            'version':this.T.version
                                        }
                                    );
                                    T.Libjs[fileName] = this.URL(u8,'text/javascript');
                                }else{
                                */
                                    contents[file.filename] = u8;
                                //}
                            }
                        }
                        return complete(key? T.Libjs[key]:contents);
                    }else{
                        return complete(data);
                    }
                };
                if(self.zip){
                    return func();
                }else{
                    return T.addJS(url,func);
                }
            });
        }

**Comment by gildas-lormeau** [comment]
I confirm that you can pass the password in the options of `ZipReader` (2nd parameter). The code looks OK in this regard.

However, the name of the function does not match the one you were debugging (`unZip` vs `unFile`).

Insert `debugger;` before `const reader = new zip.ZipReader(new zip.Uint8ArrayReader(u8),｛password｝);` and check in your debugger that `password` has the expected value, e.g. `'kk'`.

**Reply by gildas-lormeau** [reply]
If the password is OK, please post the content of the Uint8Array, i.e. `JSON.stringify(Array.from(kk.reader.array))`

**Reply by nenge123** [reply]
kk is my debug
window.kk = reader;

so you will see kk.array

-----

i not set password can see my three txt file
await unfile(kk.reader.array)


-------
console.log(kk)
can see config when my php not working 
{password: undefined }

------
when my php working 
console.log(kk)
config {passsword:'12345678'}

getData（）  output a file,filename "m.zip" is same my u8

**Reply by gildas-lormeau** [reply]
I'm sorry but I don't have enough information to help you. To know if it's a bug in zip.js, please confirm or not that the file can be opened here: https://gildas-lormeau.github.io/zip.js/demos/demo-read-file.html. 
If it's not the case then it means that zip.js might have a bug or that data are corrupted, and I need a zip file for debugging it.

If necessary, you can download the zip file of `kk.reader.array` by running the code below in the debugger:
```js
(() => {
  const link = document.createElement("a");
  link.href = URL.createObjectURL(new Blob([kk.reader.array], { type: "application/zip" }));
  link.click();
})()
```

You can also attach the file here by dropping it in the editor.

**Reply by gildas-lormeau** [reply]
I would be curious to know if the data is encrypted with AES-256 though.

**Comment by nenge123** [comment]
```javascript
let u8 = new Uint8Array([80,75,3,4,20,0,0,0,0,0,152,144,216,84,0,0,0,0,0,0,0,0,0,0,0,0,10,0,0,0,196,227,180,243,210,175,46,116,120,116,80,75,3,4,20,0,0,0,0,0,132,145,216,84,128,244,216,118,12,0,0,0,12,0,0,0,16,0,0,0,200,195,206,210,200,165,118,182,238,205,226,118,46,116,120,116,254,255,132,40,143,190,132,40,143,190,102,47,80,75,3,4,20,0,0,0,0,0,124,145,216,84,192,232,220,131,10,0,0,0,10,0,0,0,12,0,0,0,200,246,183,201,200,247,176,161,46,116,120,116,200,248,180,239,200,248,180,239,202,199,80,75,1,2,20,0,20,0,0,0,0,0,152,144,216,84,0,0,0,0,0,0,0,0,0,0,0,0,10,0,0,0,0,0,0,0,0,0,32,0,0,0,0,0,0,0,196,227,180,243,210,175,46,116,120,116,80,75,1,2,20,0,20,0,0,0,0,0,132,145,216,84,128,244,216,118,12,0,0,0,12,0,0,0,16,0,0,0,0,0,0,0,0,0,32,0,0,0,40,0,0,0,200,195,206,210,200,165,118,182,238,205,226,118,46,116,120,116,80,75,1,2,20,0,20,0,0,0,0,0,124,145,216,84,192,232,220,131,10,0,0,0,10,0,0,0,12,0,0,0,0,0,0,0,0,0,32,0,0,0,98,0,0,0,200,246,183,201,200,247,176,161,46,116,120,116,80,75,5,6,0,0,0,0,3,0,3,0,176,0,0,0,150,0,0,0,0,0]);

await DZXF.unitl.unFile(u8,null)
//{你大爷.txt: Uint8Array(0), 让我去v额外v.txt: Uint8Array(12), 撒飞洒啊.txt: Uint8Array(10)}
await DZXF.unitl.unFile(u8,null,123)
//{你大爷.txt: Uint8Array(0), 让我去v额外v.txt: Uint8Array(12), 撒飞洒啊.txt: Uint8Array(10)}
await DZXF.unitl.unFile(u8,null,undefined)
//{你大爷.txt: Uint8Array(0), 让我去v额外v.txt: Uint8Array(12), 撒飞洒啊.txt: Uint8Array(10)}
await DZXF.unitl.unFile(u8,null,undefined)
//{你大爷.txt: Uint8Array(0), 让我去v额外v.txt: Uint8Array(12), 撒飞洒啊.txt: Uint8Array(10)}
await DZXF.unitl.unFile(u8,null,"123243534")
//common_forumlocal.js?Xrh:310 
// Uncaught (in promise) DOMException: A length of 0 was specified for PBKDF2's Derive Bits operation.
await DZXF.unitl.unFile(u8,null,123574574)
//{你大爷.txt: Uint8Array(0), 让我去v额外v.txt: Uint8Array(12), 撒飞洒啊.txt: Uint8Array(10)}
await DZXF.unitl.unFile(u8,null,"123574574")
//Uncaught (in promise) DOMException: A length of 0 was specified for PBKDF2's Derive Bits operation.

```

**Comment by nenge123** [comment]
Sorry, it turns out that the clown is me! When I uploaded the file, I converted it into a zipfile because I judged that it was not a picture. As a result, my uploaded zipfile was compressed again

**Reply by gildas-lormeau** [reply]
I confirm that in your example, zip entries are not encrypted. I just fixed a minor issue in order to ignore the password in that case (see https://github.com/gildas-lormeau/zip.js/commit/ac3b092c30fab5fe87ba522a00bfbec2a8366d11). Otherwise the promise returned by `Entry#getData` was never resolved/rejected.

**Reply by nenge123** [reply]
that's great! thanks a lot.

**Comment by nenge123** [comment]
i see doc

Promise.race(iterable);

will best than the try/catch

**Reply by gildas-lormeau** [reply]
Indeed, you should never use a a classic for loop with async code (see `for(var i=0;i<entries.length;i++) {...}` in [this code](https://github.com/gildas-lormeau/zip.js/discussions/342#discussioncomment-3141526)) unless you're doing it on purpose.

**Reply by nenge123** [reply]
> Indeed, you should never use a a classic for loop with async code (see `for(var i=0;i<entries.length;i++) {...}` in [this code](https://github.com/gildas-lormeau/zip.js/discussions/342#discussioncomment-3141526)) unless you're doing it on purpose.

but i want get all file data to save/read

are you have a better than it method?

**Reply by gildas-lormeau** [reply]
You should use a `for of` loop or `Promise.all`. See the code below for example.

https://github.com/gildas-lormeau/zip.js/blob/9e65bce1557483154fe576319f4340898a858e93/tests/all/test-parallel-reads.js#L23-L26

**Reply by nenge123** [reply]
thanks a lot

**Reply by gildas-lormeau** [reply]
You're welcome! By the way, if you can I would recommend using https://www.deepl.com/translator. It gives very good results.

**Comment by nenge123** [comment]
Still having issues
password:int 123456789
password:string 123456789
int can't using .must toString();

Thank you at last, the password problem that was plaguing me was solved.

```javascript
        async unZip (u8,cb,password,key){
            let T=this.T;
            await this.ZipInitJS();
            let entries = await (new zip.ZipReader(new zip.Uint8ArrayReader(u8))).getEntries();
            if(entries.length>0){
                let contents = {};
                await Promise.all(
                    entries.map(
                        async entry=>{
                            if(!entry.directory)contents[entry.filename] = await this.ZipReadEntry(entry,password);
                            return true;
                        }
                    )
                );
                if(key){
                    Object.assign(T.Libjs,contents);
                }
                return key? T.Libjs[key]:contents;
            }else{
                return u8;
            }
        }
        async ZipReadEntry(entry,password){
            if(!entry.encrypted)return await entry.getData(new zip.Uint8ArrayWriter());
            else{
                try{
                    return await entry.getData(new zip.Uint8ArrayWriter(),{password});
                }catch(e){
                    if(e.message=='File contains encrypted entry'){
                        let newpassword = window.prompt('need a read password');
                        if(newpassword){
                            return this.ZipReadEntry(entry,newpassword);
                        }
                    }
                    throw 'miss password';
                }
            }
        }
```

**Comment by nenge123** [comment]
Still having issues
password:int 123456789
password:string 123456789
int can't using .must toString();

Thank you at last, the password problem that was plaguing me was solved.

```javascript
        async unZip (u8,cb,password,key){
            let T=this.T;
            await this.ZipInitJS();
            let entries = await (new zip.ZipReader(new zip.Uint8ArrayReader(u8))).getEntries();
            if(entries.length>0){
                let contents = {};
                await Promise.all(
                    entries.map(
                        async entry=>{
                            if(!entry.directory)contents[entry.filename] = await this.ZipReadEntry(entry,password);
                            return true;
                        }
                    )
                );
                if(key){
                    Object.assign(T.Libjs,contents);
                }
                return key? T.Libjs[key]:contents;
            }else{
                return u8;
            }
        }
        async ZipReadEntry(entry,password){
            if(!entry.encrypted)return await entry.getData(new zip.Uint8ArrayWriter());
            else{
                try{
                    return await entry.getData(new zip.Uint8ArrayWriter(),{password});
                }catch(e){
                    if(e.message=='File contains encrypted entry'){
                        let newpassword = window.prompt('need a read password');
                        if(newpassword){
                            return this.ZipReadEntry(entry,newpassword);
                        }
                    }
                    throw 'miss password';
                }
            }
        }
```

**Reply by gildas-lormeau** [reply]
Please update zip.js and please describe the issue, if any. You didn't provide any useful information. I need at least a stack trace or an error message...
